Rendering/OpenGL2/vtkOpenGLShaderProperty.h

```diff
@@ -0,0 +1,138 @@
+     * added or removed. This timestamp can be used by mappers to
+     * determine if the shader must be recompiled. Simply changing
+     * the value of an existing uniform variable doesn't affect this
+     * timestamp as it doesn't change the shader code.
+     * @return timestamp of the last modificatio
+     */
+    vtkMTimeType GetShaderMTime();
+
+    virtual void AddVertexShaderReplacement(
+      const std::string& originalValue,
+      bool replaceFirst,  // do this replacement before the default
+      const std::string& replacementValue,
+      bool replaceAll) override;
+    virtual void AddFragmentShaderReplacement(
+      const std::string& originalValue,
+      bool replaceFirst,  // do this replacement before the default
+      const std::string& replacementValue,
+      bool replaceAll) override;
+
+    virtual int GetNumberOfShaderReplacements() override;
+    virtual std::string GetNthShaderReplacementTypeAsString( vtkIdType index ) override;
+    virtual void GetNthShaderReplacement(
+      vtkIdType index,
+      std::string & name,
+      bool & replaceFirst,
+      std::string & replacementValue,
+      bool & replaceAll ) override;
+
+    virtual void ClearVertexShaderReplacement(
+      const std::string& originalValue,
+      bool replaceFirst) override;
+    virtual void ClearFragmentShaderReplacement(
+      const std::string& originalValue,
+      bool replaceFirst) override;
+    virtual void ClearAllVertexShaderReplacements() override;
+    virtual void ClearAllFragmentShaderReplacements() override;
+    virtual void ClearAllGeometryShaderReplacements() override;
+    virtual void ClearAllShaderReplacements() override;
+
+    //@{
+    /**
+     * This function enables you to apply your own substitutions
+     * to the shader creation process. The shader code in this class
+     * is created by applying a bunch of string replacements to a
+     * shader template. Using this function you can apply your
+     * own string replacements to add features you desire.
+     */
+    void AddShaderReplacement(
+      vtkShader::Type shaderType, // vertex, fragment, etc
+      const std::string& originalValue,
+      bool replaceFirst,  // do this replacement before the default
+      const std::string& replacementValue,
+      bool replaceAll);
+    void ClearShaderReplacement(
+      vtkShader::Type shaderType, // vertex, fragment, etc
+      const std::string& originalValue,
+      bool replaceFirst);
+    void ClearAllShaderReplacements(vtkShader::Type shaderType);
+    //@}
+
+    /**
+     * @brief GetAllShaderReplacements returns all user-specified shader
+     * replacements. It is provided for iteration purpuses only (const)
+     * and is mainly used by mappers when building the shaders.
+     * @return const reference to internal map holding all replacements
+     */
+    typedef std::map<const vtkShader::ReplacementSpec, vtkShader::ReplacementValue> ReplacementMap;
```

Comment:
Yes, in theory. The difficulty is that the base class is part of vtkRenderingCore (so that vtkShaderProperty can be used by vtkVolume), but vtkShader and vtkShaderProgram are defined in vtkRenderingOpenGL2 and of course, vtkRenderingCore shouldn't need to know anything about vtkRenderingOpenGL2. 

Another way to go would be to split vtkShader and vtkShaderProgram into base classes that do not know anything about OpenGL and that live in vtkRenderingCore and OpenGL implementation classes (vtkOpenGLShader and vtkOpenGLShaderProgram) that live in vtkRenderingOpenGL2 but I decided not to go there.

Reply:
Seems reasonable. I don't think it makes sense to generalize it too much as this will probably only be used with OpenGL I guess.